### ISSUE TITLE
using Sockets= in a drop-in for instantiated service doesn't clear the list of previously defined sockets

### ISSUE THREAD

**Issue by nicki-krizek** [systemd/systemd#12415]
**systemd version the issue has been seen with**
> 242

**Used distribution**
 > Archlinux

**Expected behaviour you didn't see**
 > when using `Sockets=` in a drop-in file for instantiated service, it should clear the list of previously defined sockets

**Unexpected behaviour you saw**
 > unlike other values which behave as a list (e.g. `Documentation=`), using `Sockets=` doesn't actually clear the list of previously defined values

**Steps to reproduce the problem**
```
$ systemctl cat kresd@doh
# /usr/lib/systemd/system/kresd@.service
[Unit]
Description=Knot Resolver daemon
Documentation=man:kresd.systemd(7)
Documentation=man:kresd(8)

[Service]
Type=notify
WorkingDirectory=/var/cache/knot-resolver
ExecStart=/usr/sbin/kresd --config=/etc/knot-resolver/kresd.conf
User=knot-resolver
Restart=on-abnormal
Sockets=kresd.socket
Sockets=kresd-tls.socket
Sockets=kresd-control@%i.socket

[Install]
WantedBy=kresd.target

# /usr/lib/systemd/system/kresd@.service.d/module-http.conf
[Service]
Sockets=kresd-doh.socket
Sockets=kresd-webmgmt.socket

# /etc/systemd/system/kresd@doh.service.d/override.conf
[Unit]
Documentation=
Documentation=man:test

[Service]
Sockets=
Sockets=kresd-doh.socket
```

Note `Sockets=` and `Documentation=` are both listed in the last lexicographically-ordered drop-in file. However, all sockets are passed to `kresd@doh.service` anyway - see below.

```
$ systemctl start kresd@doh
$ systemctl list-sockets kresd@doh
0 sockets listed.
Pass --all to see loaded but inactive sockets, too.
[root@localhost ~]# systemctl list-sockets kresd*
LISTEN                         UNIT                     ACTIVATES
/run/knot-resolver/control@doh kresd-control@doh.socket kresd@doh.service
127.0.0.1:44353                kresd-doh.socket         kresd@1.service, kresd@doh.service
127.0.0.1:53                   kresd.socket             kresd@1.service, kresd@doh.service
127.0.0.1:53                   kresd.socket             kresd@1.service, kresd@doh.service
127.0.0.1:8453                 kresd-webmgmt.socket     kresd@1.service, kresd@doh.service
127.0.0.1:853                  kresd-tls.socket         kresd@1.service, kresd@doh.service
[::1]:44353                    kresd-doh.socket         kresd@1.service, kresd@doh.service
[::1]:53                       kresd.socket             kresd@1.service, kresd@doh.service
[::1]:53                       kresd.socket             kresd@1.service, kresd@doh.service
[::1]:8453                     kresd-webmgmt.socket     kresd@1.service, kresd@doh.service
[::1]:853                      kresd-tls.socket         kresd@1.service, kresd@doh.service
[::]:443                       kresd-doh.socket         kresd@1.service, kresd@doh.service
```

The drop-in file was loaded correctly, and `Documentation=` was overriden:

```
$ systemctl status kresd@doh
● kresd@doh.service - Knot Resolver daemon
   Loaded: loaded (/usr/lib/systemd/system/kresd@.service; disabled; vendor preset: disabled)
  Drop-In: /usr/lib/systemd/system/kresd@.service.d
           └─module-http.conf
           /etc/systemd/system/kresd@doh.service.d
           └─override.conf
   Active: active (running) since Fri 2019-04-26 13:28:57 UTC; 12s ago
     Docs: man:test
 Main PID: 590 (kresd)
    Tasks: 1 (limit: 2359)
   Memory: 19.7M
   CGroup: /system.slice/system-kresd.slice/kresd@doh.service
           └─590 /usr/sbin/kresd --config=/etc/knot-resolver/kresd.conf
```

The sockets themselves don't explicitly require `kresd@doh.service`, only `kresd@1.service`:

```
$ systemctl cat kresd.socket
# /usr/lib/systemd/system/kresd.socket
[Unit]
Description=Knot Resolver DNS socket
Documentation=man:kresd.systemd(7)
Documentation=man:kresd(8)
Before=sockets.target

[Socket]
FreeBind=true
BindIPv6Only=both
FileDescriptorName=dns
ListenDatagram=[::1]:53
ListenStream=[::1]:53
ListenDatagram=127.0.0.1:53
ListenStream=127.0.0.1:53
Service=kresd@1.service
Slice=system-kresd.slice

[Install]
WantedBy=sockets.target
```

If I understand everything correctly, `kresd@doh.service` should not be passed sockets other than `kresd-doh.socket`.

**Comment by poettering** [comment]
Sockets= is a form of dependency (or actually just implies dependencies), and dependencies cannot be undone once they are in place. We should really document that better...

**Comment by tomaskrizek** [comment]
Currently, `systemd.service(5)` states:

```
Sockets=
...
This option may appear more than once, in which case the list of socket 
units is merged. If the empty string is assigned to this option, the list 
of sockets is reset, and all prior uses of this setting will have no effect.
```

**Comment by poettering** [comment]
urks, that's wrong actually...

**Comment by poettering** [comment]
Fix waiting in #12868

**Comment by dkg** [comment]
So this fixed the documentation, but not the underlying issue.  How do you recommend that an administrator disable a specific socket for a socket-activated service that is configured in this way?
